data/tutorials/language/0lg_00_basic_datatypes.md

```diff
@@ -0,0 +1,844 @@
+
+In the end, the only type construction that does not reduce to a variant is the function arrow type. It is possible to perform pattern matching on values of any type, except functions.
+
+#### User-Defined Polymorphic
+
+Here is an example of a variant type that combines constructors with data, constructors without data, polymorphism, and recursion:
+```ocaml
+# type 'a tree =
+  | Leaf
+  | Node of 'a * 'a tree * 'a tree;;
+type 'a tree = Leaf | Node of 'a * 'a tree * 'a tree
+```
+
+It can be used to represent arbitrarily labelled binary trees. Assuming such a tree would be labelled with integers, here is a possible way to compute the sum of the integers it contains, using recursion and pattern-matching.
+```ocaml
+# let rec sum = function
+  | Leaf -> 0
+  | Node (x, lft, rht) -> x + sum lft + sum rht;;
+val sum : int tree -> int = <fun>
+```
+
+Here is how the map function can be defined in this type:
+```ocaml
+# let rec map f = function
+  | Leaf -> Leaf
+  | Node (x, lft, rht) -> Node (f x, map f lft, map f rht);;
+val map : ('a -> 'b) -> 'a tree -> 'b tree = <fun>
+```
+
+In the OCaml community as well as in the larger functional programming community, the word *polymorphism* is used loosely. It is applied to things working in a similar fashion with various types. Several features of OCaml are polymorphic, in the broad sense, each is using a particular form of it, including its own name. In summary, OCaml has several forms of polymorphism. In most cases, the distinction between those concepts is blurred, but it sometimes needs to distinguish them. Here are the terms applicable to datatypes.
+1. `'a list`, `'a option`, and `'a tree` are very often said to be polymorphic types. Formally, `bool list` or `int option` are the types, whilst `list` and `option` are [type operators](https://en.wikipedia.org/wiki/Type_constructor) that take a type as a parameter and result in a type. This is a form of [parametric polymorphism](https://en.wikipedia.org/wiki/Parametric_polymorphism). `'a list` and `'a option` denote [type families](https://en.wikipedia.org/wiki/Type_family), which are all the types created by applying type parameters to the operators.
+2. OCaml has something called _Polymorphic Variants_. Although the types `option`, `list`, and `tree` are variants and polymorphic, they aren't polymorphic variants. They are type-parametrised variants. We stick to this usage and say the variants in this section are polymorphic. OCaml polymorphic variants are [covered in another tutorial](https://v2.ocaml.org/learn/tutorials/labels.html).
+
+### Records
+
+Records are like tuples in that several values are bundled together. In a tuple, elements are identified by their position in the corresponding product type. They are either first, second, third, or at some other position. In a record, each element has a name, it's a field. That's why record types must be declared before being used.
```

Comment:
Let's clarify what this sentence means: "In a record, each element has a name, it's a field."

Each element has a name, and that name is a field? The element is a field? There is a field for the name?

Reply:
(I changed it to the first sentence, but we can further revise it if necessary.)

Reply:
> In a record, each element has a name and a value. This name-value pair is known as a field.

This is the best wording